Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -86,240 +90,769 @@ local macro "pderiv_simp" : tactic =>
+
+lemma dblX_smul (P : Fin 3 → R) (u : Rˣ) : V.dblX (u • P) = (u ^ 4) ^ 2 * V.dblX P := by
+  simp only [dblX, negY_smul, smul_fin3_ext]
+  ring1
+
+lemma dblX_of_Z_eq_zero {P : Fin 3 → R} (hP : V.Equation P) (hPz : P z = 0) :
+    V.dblX P = (P x ^ 2) ^ 2 := by
+  linear_combination (norm := (rw [dblX, negY_of_Z_eq_zero hPz, hPz]; ring1))
+    -8 * P x * (equation_of_Z_eq_zero hPz).mp hP
+
+lemma dblX_of_Y_eq [NoZeroDivisors R] {P Q : Fin 3 → R} (hQz : Q z ≠ 0)
+    (hx : P x * Q z ^ 2 = Q x * P z ^ 2) (hy : P y * Q z ^ 3 = Q y * P z ^ 3)
+    (hy' : P y * Q z ^ 3 = V.negY Q * P z ^ 3) : V.dblX P = eval P V.polynomialX ^ 2 := by
+  rw [dblX, eval_polynomialX, Y_eq_negY_of_Y_eq hQz hx hy hy']
+  ring1
+
+private lemma toAffine_addX_of_eq {P : Fin 3 → F} {n d : F} (hPz : P z ≠ 0) (hd : d ≠ 0) :
+    W.toAffine.addX (P x / P z ^ 2) (P x / P z ^ 2) (n / (P z * d)) =
+      (n ^ 2 + W.a₁ * n * P z * d - W.a₂ * P z ^ 2 * d ^ 2 - 2 * P x * d ^ 2) / (P z * d) ^ 2 := by
+  field_simp [mul_ne_zero hPz hd]
+  ring1
+
+lemma dblX_of_Z_ne_zero {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation Q) (hPz : P z ≠ 0)
+    (hQz : Q z ≠ 0) (hx : P x * Q z ^ 2 = Q x * P z ^ 2) (hy : P y * Q z ^ 3 ≠ W.negY Q * P z ^ 3) :
+    W.dblX P / W.dblZ P ^ 2 = W.toAffine.addX (P x / P z ^ 2) (Q x / Q z ^ 2)
+      (W.toAffine.slope (P x / P z ^ 2) (Q x / Q z ^ 2) (P y / P z ^ 3) (Q y / Q z ^ 3)) := by
+  rw [dblX, toAffine_slope_of_eq hP hQ hPz hQz hx hy, dblZ,
+    ← (div_eq_div_iff (pow_ne_zero 2 hPz) (pow_ne_zero 2 hQz)).mpr hx,
+    toAffine_addX_of_eq hPz <| sub_ne_zero_of_ne <| Y_ne_negY_of_Y_ne' hP hQ hx hy]
+
+variable (V) in
+/-- The $Y$-coordinate of the doubling of a point representative, before applying the final negation
+that maps $Y$ to $-Y - a_1XZ - a_3Z^3$. -/
+def dblY' (P : Fin 3 → R) : R :=
+  (3 * P x ^ 2 + 2 * V.a₂ * P x * P z ^ 2 + V.a₄ * P z ^ 4 - V.a₁ * P y * P z)
+      * (V.dblX P - P x * (P y - V.negY P) ^ 2) + P y * (P y - V.negY P) ^ 3
+
+lemma dblY'_smul (P : Fin 3 → R) (u : Rˣ) : V.dblY' (u • P) = (u ^ 4) ^ 3 * V.dblY' P := by
+  simp only [dblY', dblX_smul, negY_smul, smul_fin3_ext]
+  ring1
+
+lemma dblY'_of_Z_eq_zero {P : Fin 3 → R} (hP : V.Equation P) (hPz : P z = 0) :
+    V.dblY' P = -(P x ^ 2) ^ 3 := by
+  linear_combination
+    (norm := (rw [dblY', dblX_of_Z_eq_zero hP hPz, negY_of_Z_eq_zero hPz, hPz]; ring1))
+    (8 * (equation_of_Z_eq_zero hPz).mp hP - 12 * P x ^ 3) * (equation_of_Z_eq_zero hPz).mp hP
+
+lemma dblY'_of_Y_eq [NoZeroDivisors R] {P Q : Fin 3 → R} (hQz : Q z ≠ 0)
+    (hx : P x * Q z ^ 2 = Q x * P z ^ 2) (hy : P y * Q z ^ 3 = Q y * P z ^ 3)
+    (hy' : P y * Q z ^ 3 = V.negY Q * P z ^ 3) : V.dblY' P = (-eval P V.polynomialX) ^ 3 := by
+  rw [dblY', dblX_of_Y_eq hQz hx hy hy', eval_polynomialX, Y_eq_negY_of_Y_eq hQz hx hy hy']
+  ring1
+
+private lemma toAffine_addY'_of_eq {P : Fin 3 → F} {n d : F} (hPz : P z ≠ 0) (hd : d ≠ 0) :
```

Comment:
this could have been `negAddY` . I think I far prefer my suggestions to just "add a ' to indicate a negation" -- usually adding a ' indicates a variant, not a completely different function.

Reply:
Ditto